curvefs/src/client/rpcclient/metacache.cpp

```diff
@@ -0,0 +1,348 @@
+    CopysetInfo<MetaserverID> *toupdateCopyset) {
+    MetaserverID metaserverID = 0;
+    PeerAddr leaderAddr;
+
+    // get leader from peers
+    bool getLeaderOk = cli2Client_->GetLeader(
+        poolID, copysetID, toupdateCopyset->csinfos_,
+        toupdateCopyset->leaderindex_, &leaderAddr, &metaserverID);
+    if (false == getLeaderOk) {
+        LOG(WARNING) << "get leader failed!"
+                     << ", copyset id = " << copysetID
+                     << ", logicpool id = " << poolID;
+        return false;
+    }
+
+    // update leader info, if fail, update copyset info from mds
+    int ret = toupdateCopyset->UpdateLeaderInfo(leaderAddr);
+    if (ret == -1 && !leaderAddr.IsEmpty()) {
+        CopysetPeerInfo<MetaserverID> metaserverInfo;
+        bool ret = mdsClient_->GetMetaServerInfo(leaderAddr, &metaserverInfo);
+        if (false == ret) {
+            LOG(ERROR) << "get metaserver id from mds failed, addr = "
+                       << leaderAddr.ToString();
+            return false;
+        }
+
+        UpdateCopysetInfoIfMatchCurrentLeader(poolID, copysetID, leaderAddr);
+        GetCopysetInfowithCopySetID(poolID, copysetID, toupdateCopyset);
+        ret = toupdateCopyset->UpdateLeaderInfo(leaderAddr, metaserverInfo);
+    }
+
+    return true;
+}
+
+void MetaCache::UpdateCopysetInfo(LogicPoolID poolID, CopysetID copysetID,
+                                  const CopysetInfo<MetaserverID> &csinfo) {
+    const auto key =
+        ::curve::client::MetaCache::CalcLogicPoolCopysetID(poolID, copysetID);
+    WriteLockGuard wl(rwlock4copysetInfoMap_);
+    copysetInfoMap_[key] = csinfo;
+}
+
+void MetaCache::UpdatePartitionInfo(uint32_t fsID,
+                                    const PatitionInfoList &pInfoList) {
+    WriteLockGuard wl(rwlock4fs2PartitionInfoMap_);
+    fs2PartitionInfoMap_[fsID] = pInfoList;
+}
+
+
+bool MetaCache::GetParitionListWihFsID(uint32_t fsID,
+                                       PatitionInfoList *pinfoList) {
+    ReadLockGuard rl(rwlock4fs2PartitionInfoMap_);
+    auto iter = fs2PartitionInfoMap_.find(fsID);
+    if (iter == fs2PartitionInfoMap_.end()) {
+        return false;
+    }
+    *pinfoList = iter->second;
+    if (pinfoList->empty()) {
+        return false;
+    }
+    return true;
+}
+
+bool MetaCache::GetCopysetIDwithInodeID(const PatitionInfoList &pinfoList,
+                                        uint32_t inodeID, LogicPoolID *poolID,
+                                        CopysetID *copysetID,
+                                        uint32_t *partitionID) {
+    bool find = false;
+    for_each(pinfoList.begin(), pinfoList.end(),
+             [&](const PartitionInfo &pinfo) {
+                 if (pinfo.start() <= inodeID && pinfo.end() >= inodeID) {
+                     find = true;
+                     *poolID = pinfo.poolid();
+                     *copysetID = pinfo.poolid();
+                     *partitionID = pinfo.partitionid();
```

Comment:
forget break when found, and confirm whether `end` is inclusive or exclusive ?

Reply:
fix. I verified [start, end].